go/mcap/indexed_message_iterator.go

```diff
@@ -191,50 +281,99 @@ func (it *indexedMessageIterator) loadChunk(chunkIndex *ChunkIndex) error {
-	var recordLen uint64
-	offset := 0
-	for offset < len(messageIndexSection) {
-		if op := OpCode(messageIndexSection[offset]); op != OpMessageIndex {
-			return fmt.Errorf("unexpected token %s in message index section", op)
-		}
-		offset++
-		recordLen, offset, err = getUint64(messageIndexSection, offset)
-		if err != nil {
-			return fmt.Errorf("failed to get message index record length: %w", err)
+	// produce message indexes for the newly decompressed chunk data.
+	chunkIsOrdered := true
+	var maxLogTime uint64
+	it.messageIndexes = it.messageIndexes[:0]
+	for offset := uint64(0); offset < bufSize; {
+		if bufSize < offset+1+8 {
+			return fmt.Errorf("expected another record in chunk, but left with %d bytes", bufSize-offset)
 		}
-		messageIndex, err := ParseMessageIndex(messageIndexSection[offset : uint64(offset)+recordLen])
-		if err != nil {
-			return fmt.Errorf("failed to parse message index: %w", err)
+		opcodeAndLengthBuf := chunkSlot.buf[offset : offset+1+8]
+		op := OpCode(opcodeAndLengthBuf[0])
+		recordLen := binary.LittleEndian.Uint64(opcodeAndLengthBuf[1:])
+		recordStart := offset + 1 + 8
+		recordEnd, overflow := checkedAdd(recordStart, recordLen)
+		if overflow {
+			return fmt.Errorf("record length extends past uint64 range: start: %d, len: %d", recordStart, recordLen)
 		}
-		offset += int(recordLen)
-		// skip message indexes for channels we don't need
-		if _, ok := it.channels[messageIndex.ChannelID]; !ok {
-			continue
+		if bufSize < recordEnd {
+			return fmt.Errorf(
+				"%s record in chunk has length %d bytes but only %d remaining in chunk",
+				op.String(), recordLen, bufSize-recordStart)
 		}
-		// push any message index entries in the requested time range to the heap to read.
-		for i := range messageIndex.Records {
-			timestamp := messageIndex.Records[i].Timestamp
-			if timestamp >= it.start && timestamp < it.end {
-				if err := it.indexHeap.HeapPush(rangeIndex{
-					chunkIndex:        chunkIndex,
-					messageIndexEntry: &messageIndex.Records[i],
-					buf:               chunkData,
-				}); err != nil {
-					return err
+		recordContent := chunkSlot.buf[recordStart:recordEnd]
+		msg := Message{}
+		switch op {
+		case OpMessage:
+			if err := msg.PopulateFrom(recordContent); err != nil {
+				return fmt.Errorf("could not parse message in chunk: %w", err)
+			}
+			if slicemap.GetAt(it.channels, msg.ChannelID) != nil {
+				if msg.LogTime >= it.start && msg.LogTime < it.end {
+					it.messageIndexes = append(it.messageIndexes, MessageIndexEntry{Timestamp: msg.LogTime, Offset: offset})
+					if msg.LogTime < maxLogTime {
+						chunkIsOrdered = false
+					} else {
+						maxLogTime = msg.LogTime
+					}
+					chunkSlot.unreadMessages++
 				}
 			}
+		case OpChannel, OpSchema:
+			// no-op
+		default:
+			return fmt.Errorf(
+				"expected only schema, channel, message opcodes in chunk, found %s at offset %d",
```

Comment:
Since the spec tells the user they are free to implement their own record types, should we be returning error here or skipping over unrecognized stuff?

Reply:
The spec is less friendly towards arbitrary records in chunks: https://mcap.dev/spec#chunk-op0x06

> A Chunk contains a batch of Schema, Channel, and Message records.